File: cslavb/Csla/Core/MobileDictionary.vb
Imports System.Collections.Generic
Imports System.Linq
Imports Csla.Serialization
Imports Csla.Serialization.Mobile

Namespace Core

  ''' <summary>
  ''' Defines a dictionary that can be serialized through
  ''' the MobileFormatter.
  ''' </summary>
  ''' <typeparam name="K">Key value: any primitive or IMobileObject type.</typeparam>
  ''' <typeparam name="V">Value: any primitive or IMobileObject type.</typeparam>
  <Serializable()> _
  Public Class MobileDictionary(Of K, V)
    Inherits Dictionary(Of K, V)
    Implements IMobileObject
    Private _keyIsMobile As Boolean
    Private _valueIsMobile As Boolean

    ''' <summary>
    ''' Creates an instance of the object.
    ''' </summary>
    Public Sub New()
      DetermineTypes()
    End Sub

    ''' <summary>
    ''' Creates an instance of the object based
    ''' on the supplied dictionary, whose elements
    ''' are copied to the new dictionary.
    ''' </summary>
    ''' <param name="capacity">The initial number of elements 
    ''' the dictionary can contain.</param>
    Public Sub New(ByVal capacity As Integer)
      MyBase.New(capacity)
      DetermineTypes()
    End Sub

    ''' <summary>
    ''' Creates an instance of the object based
    ''' on the supplied dictionary, whose elements
    ''' are copied to the new dictionary.
    ''' </summary>
    ''' <param name="comparer">The comparer to use when
    ''' comparing keys.</param>
    Public Sub New(ByVal comparer As IEqualityComparer(Of K))
      MyBase.New(comparer)
      DetermineTypes()
    End Sub

    ''' <summary>
    ''' Creates an instance of the object based
    ''' on the supplied dictionary, whose elements
    ''' are copied to the new dictionary.
    ''' </summary>
    ''' <param name="dict">Source dictionary.</param>
    Public Sub New(ByVal dict As Dictionary(Of K, V))
      MyBase.New(dict)
      DetermineTypes()
    End Sub

    Private Sub DetermineTypes()
      _keyIsMobile = GetType(Csla.Serialization.Mobile.IMobileObject).IsAssignableFrom(GetType(K))
      _valueIsMobile = GetType(Csla.Serialization.Mobile.IMobileObject).IsAssignableFrom(GetType(V))
    End Sub

#Region "IMobileObject Members"

    Private Const _keyPrefix As String = "k"
    Private Const _valuePrefix As String = "v"

    Private Sub GetState(ByVal info As SerializationInfo) Implements IMobileObject.GetState
      info.AddValue("count", Me.Keys.Count)
    End Sub

    Private Sub GetChildren(ByVal info As SerializationInfo, ByVal formatter As MobileFormatter) Implements IMobileObject.GetChildren
      Dim count As Integer = 0
      For Each key In Me.Keys
        If _keyIsMobile Then
          Dim si As SerializationInfo = formatter.SerializeObject(key)
          info.AddChild(_keyPrefix & count, si.ReferenceId)
        Else
          info.AddValue(_keyPrefix & count, key)
        End If

        If _valueIsMobile Then
          Dim si As SerializationInfo = formatter.SerializeObject(Me(key))
          info.AddChild(_valuePrefix & count, si.ReferenceId)
        Else
          Dim value As V = Me(key)
          info.AddValue(_valuePrefix & count, value)
        End If
        count += 1
      Next key
    End Sub

    Private Sub SetState(ByVal info As SerializationInfo) Implements IMobileObject.SetState
    End Sub

    Private Sub SetChildren(ByVal info As SerializationInfo, ByVal formatter As MobileFormatter) Implements IMobileObject.SetChildren
      Dim count As Integer = info.GetValue(Of Integer)("count")

      For index As Integer = 0 To count - 1
        Dim key As K
        If _keyIsMobile Then
          key = CType(formatter.GetObject(info.Children(_keyPrefix & index).ReferenceId), K)
        Else
          key = info.GetValue(Of K)(_keyPrefix & index)
        End If

        Dim value As V
        If _valueIsMobile Then
          value = CType(formatter.GetObject(info.Children(_valuePrefix & index).ReferenceId), V)
        Else
          value = info.GetValue(Of V)(_valuePrefix & index)
        End If

        Add(key, value)
      Next index
    End Sub

#End Region
  End Class

End Namespace

File: Source/Csla/Core/MobileDictionary.cs
//-----------------------------------------------------------------------
// <copyright file="MobileDictionary.cs" company="Marimer LLC">
//     Copyright (c) Marimer LLC. All rights reserved.
//     Website: https://cslanet.com
// </copyright>
// <summary>Defines a dictionary that can be serialized through</summary>
//-----------------------------------------------------------------------
using System.Collections.Generic;
using System.Linq;
using Csla.Serialization.Mobile;
using System;
using System.Reflection;
using Csla.Reflection;

namespace Csla.Core
{
  /// <summary>
  /// Defines a dictionary that can be serialized through
  /// the SerializationFormatterFactory.GetFormatter().
  /// </summary>
  /// <typeparam name="K">Key value: any primitive or IMobileObject type.</typeparam>
  /// <typeparam name="V">Value: any primitive or IMobileObject type.</typeparam>
  [Serializable()]
  public class MobileDictionary<K, V> : Dictionary<K, V>, IMobileObject
  {
    private bool _keyIsMobile;
    private bool _valueIsMobile;

    /// <summary>
    /// Creates an instance of the object.
    /// </summary>
    public MobileDictionary()
    {
      DetermineTypes();
    }

    /// <summary>
    /// Creates an instance of the object based
    /// on the supplied dictionary, whose elements
    /// are copied to the new dictionary.
    /// </summary>
    /// <param name="capacity">The initial number of elements 
    /// the dictionary can contain.</param>
    public MobileDictionary(int capacity)
      : base(capacity)
    {
      DetermineTypes();
    }

    /// <summary>
    /// Creates an instance of the object based
    /// on the supplied dictionary, whose elements
    /// are copied to the new dictionary.
    /// </summary>
    /// <param name="comparer">The comparer to use when
    /// comparing keys.</param>
    public MobileDictionary(IEqualityComparer<K> comparer)
      : base(comparer)
    {
      DetermineTypes();
    }

    /// <summary>
    /// Creates an instance of the object based
    /// on the supplied dictionary, whose elements
    /// are copied to the new dictionary.
    /// </summary>
    /// <param name="dict">Source dictionary.</param>
    public MobileDictionary(Dictionary<K, V> dict)
      : base(dict)
    {
      DetermineTypes();
    }

    /// <summary>
    /// Gets a value indicating whether the
    /// dictionary contains the specified key
    /// value.
    /// </summary>
    /// <param name="key">Key value</param>
    public bool Contains(K key)
    {
      return base.ContainsKey(key);
    }

#if !(ANDROID || IOS) && !NETFX_CORE
    /// <summary>
    /// Creates an instance of the object for serialization.
    /// </summary>
    /// <param name="context">Serialization context.</param>
    /// <param name="info">Serialization info.</param>
    protected MobileDictionary(System.Runtime.Serialization.SerializationInfo info,
        System.Runtime.Serialization.StreamingContext context)
      : base(info, context)
    {
      DetermineTypes();
    }
#endif

    private void DetermineTypes()
    {
      _keyIsMobile = typeof(Csla.Serialization.Mobile.IMobileObject).IsAssignableFrom(typeof(K));
      _valueIsMobile = typeof(Csla.Serialization.Mobile.IMobileObject).IsAssignableFrom(typeof(V));
    }

    #region IMobileObject Members

    private const string _keyPrefix = "k";
    private const string _valuePrefix = "v";

    void IMobileObject.GetState(SerializationInfo info)
    {
      info.AddValue("count", this.Keys.Count);
    }

    void IMobileObject.GetChildren(SerializationInfo info, MobileFormatter formatter)
    {
      int count = 0;
      foreach (var key in this.Keys)
      {
        if (_keyIsMobile)
        {
          SerializationInfo si = formatter.SerializeObject(key);
          info.AddChild(_keyPrefix + count, si.ReferenceId);
        }
        else
        {
          info.AddValue(_keyPrefix + count, key);
        }

        if (_valueIsMobile)
        {
          SerializationInfo si = formatter.SerializeObject(this[key]);
          info.AddChild(_valuePrefix + count, si.ReferenceId);
        }
        else
        {
          V value = this[key];
          info.AddValue(_valuePrefix + count, value);
        }
        count++;
      }
    }

    void IMobileObject.SetState(SerializationInfo info)
    { }

    void IMobileObject.SetChildren(SerializationInfo info, MobileFormatter formatter)
    {
      int count = info.GetValue<int>("count");

      for (int index = 0; index < count; index++)
      {
        K key;
        if (_keyIsMobile)
          key = (K)formatter.GetObject(info.Children[_keyPrefix + index].ReferenceId);
        else
          key = info.GetValue<K>(_keyPrefix + index);

        V value;
        if (_valueIsMobile)
          value = (V)formatter.GetObject(info.Children[_valuePrefix + index].ReferenceId);
        else
          value = info.GetValue<V>(_valuePrefix + index);

        Add(key, value);
      }
    }

    #endregion
  }
}